### PR TITLE
Try to provide some thread safety to msg_stack

### LINKED ISSUE #1511: TRACE is not thread-safe?
`TRACE` seems not to be thread-safe, and can cause segfaults if called from inside an OpenMP parallelized loop, see #1510. This is unfortunate because it means any function/method that might ever be called inside an OpenMP loop cannot use `TRACE`/`AUTO_TRACE`. I think this behaviour at least needs documenting. I don't know if there is a simple fix; could we perhaps make a separate `msg_stack` for each thread (maybe as a `std::vector` whose size is the maximum number of threads)?

### PR DESCRIPTION
**Pull request by d7919**
A better longer term fix is probably worth it but hopefully this will
help in the interim.

Attempts to fix #1511

### CODE REVIEW

File: src/sys/msg_stack.cxx [hunk 1]
```diff
@@ -24,30 +24,33 @@
...
+    if (position >= stack.size()) {
+      stack.emplace_back(buffer);
+    } else {
+      stack[position] = buffer;
+    }
 
-  return position++;
+    position++;
+  };
+  return position;
```
Comment (d7919) [inline src/sys/msg_stack.cxx]: position-1

File: src/sys/msg_stack.cxx [hunk 2]
```diff
@@ -58,25 +61,32 @@ int MsgStack::setPoint() {
...
 
-  if (id > static_cast<int>(position))
-    return;
+  BOUT_OMP(critical(MsgStack::pop)) {
+    if (id <= static_cast<int>(position))
+      position = id;
+  };
 
-  position = id;
+  return;
```
Comment (ZedThree) [inline src/sys/msg_stack.cxx]: `return` is superfluous here :)

**Comment by d7919** [comment]
The section name isn't strictly required but is strongly recommended to prevent locks.

**Comment by d7919** [comment]
@ZedThree should this be for 4.2.1 or into next?

**Comment by d7919** [comment]
There's perhaps a concern about the performance impact of these critical sections if `TRACE` is used in a tight loop -- the CI times don't seem to indicate a big change but these are not the most reliable performance indicators. It might be worth checking the performance with and without this PR for some realistic cases using significant numbers of threads?

**Comment by ZedThree** [comment]
This is an existing problem in `master`, so yes, should be in to `v4.2.1-rc` please.

It would be nice to get a feel for the performance impact, but `TRACE` is really a debugging tool and can always be turned off, so I'm not overly concerned.

**Comment by d7919** [comment]
Replaced by #1517

**Pull request closed without merging**